### PR TITLE
[REVERTED] VectorBuilder shares structure of added Vector if the builder was empty, for better performance of Vector#concat

### LINKED ISSUE #11329: VectorBuilder#addAll(Vector) should perform a "shallow copy" of the argument, when the builder is currently empty.
Take this example
```scala
val v = Vector.fill(1000000)(1)

val b = Vector.newBuilder

b.addAll(v)
```

Here, `b` will perform a deep copy of `v`, but it should be able to just do something along the lines of `this.initFrom(v)` and whatever other bookkeeping is required within the builder. 

This will be a really great improvement to performance, especially because this is how Vectors currently perform concatenation in most cases (`newBuilder.addAll(this).addAll(that).result()`)

### PR DESCRIPTION
**Pull request by joshlemer**
Fixes https://github.com/scala/bug/issues/11329

Currently, when adding a Vector to a VectorBuilder via `builder.addAll(v)`, that builder will copy all elements into the new Vector. Instead we can share almost all structure with that passed Vector, if the builder was empty before adding the Vector.

This is important because this is the code-path hit by `Vector#concat(Iterable)` and so by doing this we greatly improve CPU time of the builder in these cases, reduce memory usage, and CPU time of the Vector#concat/appendedAll method.

#### Vector#concat

The benefit to Vector concat is large but not as large as the benefit to the builder. That is because there was already some logic inside Vector concat that would avoid copying all elements of the left vector, and instead prepend the elements of `this` to the vector on the right, one-by-one. Still, that trick does not work for non-Vector collections. 

```
[info] Benchmark                            (sizeLeft)  (sizeRight)  Mode  Cnt      Score     Error  Units
[info] VectorBenchmark.leftConcatRight_new           0          100  avgt    6      6.756 ±   0.579  ns/op
[info] VectorBenchmark.leftConcatRight_new           1          100  avgt    6     49.904 ±   1.709  ns/op
[info] VectorBenchmark.leftConcatRight_new          10          100  avgt    6    218.261 ±   7.919  ns/op
[info] VectorBenchmark.leftConcatRight_new         100          100  avgt    6    318.429 ±  18.978  ns/op
[info] VectorBenchmark.leftConcatRight_new        1000          100  avgt    6    262.965 ±  12.497  ns/op
[info] VectorBenchmark.leftConcatRight_new       10000          100  avgt    6    493.761 ±   8.534  ns/op
[info] VectorBenchmark.leftConcatRight_new      100000          100  avgt    6    644.133 ±  13.492  ns/op
[info] VectorBenchmark.leftConcatRight_new     1000000          100  avgt    6    637.583 ±  27.483  ns/op
[info] VectorBenchmark.leftConcatRight_new    10000000          100  avgt    6    839.363 ±  95.754  ns/op
[info] VectorBenchmark.leftConcatRight_old           0          100  avgt    6     15.956 ±   0.437  ns/op
[info] VectorBenchmark.leftConcatRight_old           1          100  avgt    6     62.119 ±   1.686  ns/op
[info] VectorBenchmark.leftConcatRight_old          10          100  avgt    6    224.170 ±   6.856  ns/op
[info] VectorBenchmark.leftConcatRight_old         100          100  avgt    6    324.611 ±  12.725  ns/op
[info] VectorBenchmark.leftConcatRight_old        1000          100  avgt    6   1218.946 ±  39.278  ns/op
[info] VectorBenchmark.leftConcatRight_old       10000          100  avgt    6  10124.572 ± 709.876  ns/op
[info] VectorBenchmark.leftConcatRight_old      100000          100  avgt    6  10329.658 ± 709.561  ns/op
[info] VectorBenchmark.leftConcatRight_old     1000000          100  avgt    6   9955.540 ± 716.100  ns/op
[info] VectorBenchmark.leftConcatRight_old    10000000          100  avgt    6  10465.319 ± 675.380  ns/op
```
#### VectorBuilder#addAll
benchmark's `new VectorBuilder[A].addAll(vec)`
```
[info] Benchmark                                (sizeLeft)  Mode  Cnt         Score         Error  Units
[info] VectorBenchmark.vectorBuilderAddAll_new           0  avgt    6        20.314 ±       1.864  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_new           1  avgt    6        37.637 ±       0.781  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_new          10  avgt    6        42.497 ±       1.103  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_new         100  avgt    6        49.754 ±       1.258  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_new        1000  avgt    6        46.884 ±       2.006  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_new       10000  avgt    6       264.986 ±      17.862  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_new      100000  avgt    6       404.341 ±       5.669  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_new     1000000  avgt    6       410.077 ±       8.319  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_new    10000000  avgt    6       564.855 ±      29.569  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_old           0  avgt    6        19.112 ±       0.352  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_old           1  avgt    6        38.005 ±       0.944  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_old          10  avgt    6        38.558 ±       0.867  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_old         100  avgt    6       145.642 ±       5.987  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_old        1000  avgt    6      1075.496 ±      23.117  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_old       10000  avgt    6     10699.170 ±     440.156  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_old      100000  avgt    6    164777.276 ±  184712.385  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_old     1000000  avgt    6   2180308.349 ±  502256.170  ns/op
[info] VectorBenchmark.vectorBuilderAddAll_old    10000000  avgt    6  38082222.389 ± 1357357.187  ns/op
```

### CODE REVIEW

File: src/library/scala/collection/immutable/Vector.scala
Comment (joshlemer) [inline src/library/scala/collection/immutable/Vector.scala]: Remove before merging

File: src/library/scala/collection/immutable/Vector.scala
Comment (joshlemer) [inline src/library/scala/collection/immutable/Vector.scala]: Remove before merging

File: src/library/scala/collection/immutable/Vector.scala
Comment (julienrf) [inline src/library/scala/collection/immutable/Vector.scala]: That `match` expression is not exhaustive (an `IterableOnce` could be something that’s not an `Iterable` nor an `Iterator`). Do you think we could merge the `Iterable` and `Iterator` cases into a single `IterableOnce` case?

File: src/library/scala/collection/immutable/Vector.scala
Comment (julienrf) [inline src/library/scala/collection/immutable/Vector.scala:790]: Do we have tests that cover that this is still correct?

**Review by julienrf — CHANGES_REQUESTED** [review]
This is a lot of added code, but the performance benefits seem worth it. I’ve left a few remarks.

File: src/library/scala/collection/immutable/Vector.scala
Comment (szeiger) [reply src/library/scala/collection/immutable/Vector.scala]: It's not obvious why they are separate in the first place. `IterableOnce` has `knownSize`, the `Iterable` case only gets a single `Iterator` and `iterator` is a no-op for `Iterator`. Changing the `Iterable` case to `IterableOnce` should work.

File: src/library/scala/collection/immutable/Vector.scala
Comment (szeiger) [inline src/library/scala/collection/immutable/Vector.scala]: What about `thisLength == 0`?

File: src/library/scala/collection/immutable/Vector.scala
Comment (szeiger) [inline src/library/scala/collection/immutable/Vector.scala]: Do we need separate versions of `gotoPositionWritable1` and `nullSlotAndCopy`? Is there a performance penalty when modifying the original versions to pass in a possibly null array?

File: src/library/scala/collection/immutable/Vector.scala
Comment (joshlemer) [reply src/library/scala/collection/immutable/Vector.scala]: Good catch

File: src/library/scala/collection/immutable/Vector.scala
Comment (joshlemer) [reply src/library/scala/collection/immutable/Vector.scala]: That is a good idea!

File: src/library/scala/collection/immutable/Vector.scala
Comment (joshlemer) [reply src/library/scala/collection/immutable/Vector.scala:790]: Not at the moment I don't think so

**Comment by joshlemer** [comment]
After doing some more thorough property testing, sometimes we get null pointer exceptions when calling `VectorBuilder#result()`, so some debugging needs to be done here.

In particular, it seem that some combination of builder mutations cause the `display1` to be `null` even though `depth == 2`

**Comment by Ichoran** [comment]
What's the status of this @joshlemer ?  Vector code is pretty hairy, especially the dirty/clean distinction.  Do you have time to work on this as well as all the other stuff?

**Comment by joshlemer** [comment]
@Ichoran 
>Vector code is pretty hairy

Very much so!

Ihaven't implemented julien or stefan's requested changes yet, and not yet 100% confident the builder's logic is completely correct for all cases of `addAll`'d vectors where the addAll'd vector has non-standard representation (because it has some start/end offset). It has something to do with `display1` being set to `null` even when `depth == 2`.

I encountered those issues when writing property tests which would randomly generate a sequence of mutations on the builder, with randomly generated vectors that have had some amount of elements dropped. 

Some further investigation would be required to figure out where the error is. I'm not sure when I'll be able to figure it out.

**Comment by Ichoran** [comment]
@joshlemer - Do you have time to figure out the display issue (there's some method that cleans it up when it's dirty, IIRC)?

**Comment by Ichoran** [comment]
@joshlemer - Did you get the display stuff sorted out?

**Comment by joshlemer** [comment]
I investigated but didn't end up figuring it out, I'm going to have an other look though. Could use some help to be honest though if someone knows the Vector code well

**Comment by Ichoran** [comment]
All right.  I probably won't have time, but I'll keep this in mind if I get more time than I expect :)

**Comment by joshlemer** [comment]
Maybe @mkeskells is very familiar with Vector codebase

**Comment by joshlemer** [comment]
Found a breaking example:
```scala
      Vector.newBuilder[Int]
        .addAll((1 to 100000).to(Vector).drop(10000).dropRight(0))
        .result()
```

```
java.lang.NullPointerException
	at scala.collection.immutable.VectorPointer.gotoPos(Vector.scala:1065)
	at scala.collection.immutable.VectorPointer.gotoPos$(Vector.scala:1054)
	at scala.collection.immutable.Vector.gotoPos(Vector.scala:77)
	at scala.collection.immutable.VectorBuilder.result(Vector.scala:963)
	at scala.collection.immutable.VectorBuilder.result(Vector.scala:780)
```

**Comment by Ichoran** [comment]
Yeah, that's a classic symptom of treating a dirty tree as if it's clean.  (That is, the display hierarchy has the correct thing in, say, display3, but the thing in display4 is out of date and doesn't have the slot corresponding to display3 actually pointing to the thing in display 3.)

**Comment by joshlemer** [comment]
Ok that broken case is fixed

File: src/library/scala/collection/immutable/Vector.scala
Comment (joshlemer) [reply src/library/scala/collection/immutable/Vector.scala]: I have now implemented these two methods in the suggested way, with a possibly null array argument to the existing methods.

File: src/library/scala/collection/immutable/Vector.scala
Comment (joshlemer) [reply src/library/scala/collection/immutable/Vector.scala:790]: I do now have a test for this, perhaps there is a better/more thorough way to test it though.

File: test/junit/scala/collection/immutable/VectorTest.scala
Comment (joshlemer) [inline test/junit/scala/collection/immutable/VectorTest.scala]: Remove before merging, just some debugging code..

File: src/library/scala/collection/immutable/Vector.scala
Comment (joshlemer) [reply src/library/scala/collection/immutable/Vector.scala]: The two implementations have now been unified

File: src/library/scala/collection/immutable/Vector.scala
Comment (joshlemer) [reply src/library/scala/collection/immutable/Vector.scala]: Consider it special-cased!

**Comment by Ichoran** [comment]
(Note to self: don't obviously have time for this now.  Maybe next time?)

**Comment by adriaanm** [comment]
@joshlemer could you clarify the status? we're in the final week of RC1 PR merging (sorry for all the pings)

**Comment by joshlemer** [comment]
@adriaanm I have now removed any benchmark code / debugging code, and so far as I can tell this implementation is correct but I think that if someone knows the vector codebase well, it would be best to get their second opinion.

**Comment by joshlemer** [comment]
An that is ok about the pings, I deserved it :-P

**Review by retronym — APPROVED** [review]

**Comment by SethTisue** [comment]
reverted: https://github.com/scala/scala/pull/8410

**Pull request merged**
